spices/SPICE-0014-canonical-formatter.adoc

```diff
@@ -0,0 +1,780 @@
+
+amends
+ "bar.pkl"
+----
+
+.output
+[source, pkl]
+----
+module foo.bar.baz
+amends "bar.pkl"
+----
+
+==== [0005] Imports
+
+Imports should be sorted alphabetically. +
+Each import should be written in single line. +
+They should be grouped in 3 groups, in this order: absolute imports, relative imports, and project imports. +
+Each group should be separated by an empty line. +
+Imports in the same group should not be separated by an empty line.
+
+.input
+[source, pkl]
+----
+import
+ "@foo/Foo.pkl"
+  as foo
+import* "**.pkl"
+
+import "package://example.com/myPackage@1.0.0#/Qux.pkl"
+
+
+import "https://example.com/baz.pkl"
+import "..."
+import "@bar/Bar.pkl"
+----
+
+.output
+[source, pkl]
+----
+import "https://example.com/baz.pkl"
+import "package://example.com/myPackage@1.0.0#/Qux.pkl"
+
+import* "**.pkl"
+import "..."
+
+import "@bar/Bar.pkl"
+import "@foo/Foo.pkl" as foo
+----
+
+==== [0006] Module and class members
+
+Module and class members must be separated by exactly one empty line.
+
+.input
+[source, pkl]
+----
+foo = 4
+
+
+bar: String = "bar"
+class Foo {
+  prop: Int; prop2: Boolean
+}
+----
+
+.output
+[source, pkl]
+----
+foo = 4
+
+bar: String = "bar"
+
+class Foo {
+  prop: Int
+
+  prop2: Boolean
+}
+----
+
+==== [0007] Members within braces
+
+Members within braces must be indented one level deeper than their parents.
+
+.input
+[source, pkl]
+----
+foo {
+bar {
+    baz = "hi"
+}
+}
+----
+
+.output
+[source, pkl]
+----
+foo {
+  bar {
+    baz = "hi"
+  }
+}
+----
+
+==== [0008] Modifiers
+
+Modifiers must be written in the same line as what they are modifying, except if they go over the maximum line limit. +
+Modifiers must be ordered alphabetically.
+
+.input
+[source, pkl]
+----
+hidden
+const
+foo = 1
+
+open
+local class Foo {}
+----
+
+.output
+[source, pkl]
+----
+const hidden foo = 1
+
+local open class Foo {}
+----
+
+==== [0009] Class header
+
+The class header must be written in a single line if it doesn't go over the maximum line limit.
+
+.input
+[source, pkl]
+----
+local
+open
+class Foo
+extends Bar {
+}
+
+class ReaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaalyLongClassName extends Bar {}
+----
+
+.output
+[source, pkl]
+----
+local open class Foo extends Bar {}
+
+class ReaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaalyLongClassName
+  extends Bar {}
+----
+
+==== [0010] Object and class bodies
+
+Object and class bodies must have the opening `{` in the same line as the previous token. +
+A new line must follow the opening `{`. +
+The closing `}` should be in their own line. +
+An empty body should have no new lines or spaces between `{` and `}`.
+
+.input
+[source, pkl]
+----
+foo = (parent) {
+
+}
+
+class Bar
+{
+  qux = 1
+}
+
+bar = new { prop = 0 }
+----
+
+.output
+[source, pkl]
+----
+foo = (parent) {}
+
+class Bar {
+  qux = 1
+}
+
+bar = new {
+  prop = 0
+}
+----
+
+==== [0011] Object members
+
+Object members must be defined in their own line. +
```

Comment:
Similarly, it'd be nice to allow "inline" properties/elements (up to two), in accordance with the style guide.